cloudsync_gdrive.py

```diff
@@ -312,56 +326,103 @@ def events(self) -> Generator[Event, None, None]:  # pylint: disable=too-many-lo
+                otype = DIRECTORY
+            else:
+                otype = FILE
+        else:
+            otype = NOTKNOWN
 
-                remove = []
-                for cpath, coid in self._ids.items():
-                    if coid == oid:
-                        if cpath != path:
-                            remove.append(cpath)  # remove the event's item if it's path changed
+        ohash = None
+        path = self._path_oid(oid, use_cache=False)
+        if not path:
+            # Trashed
+            exists = False
 
-                    if path and otype == DIRECTORY and self.is_subpath(path, cpath):
-                        remove.append(cpath)  # if the event's item is a folder, uncache its children
+        event = Event(otype, oid, path, ohash, exists, ts, new_cursor=new_cursor)
 
-                for r in remove:
-                    self._ids.pop(r, None)
+        remove = []
+        for cpath, coid in self._ids.items():
+            if coid == oid:
+                if cpath != path:
+                    remove.append(cpath)  # remove the event's item if it's path changed
 
-                if path:
-                    self._ids[path] = oid
+            if path and otype == DIRECTORY and self.is_subpath(path, cpath):
+                remove.append(cpath)  # if the event's item is a folder, uncache its children
 
-                log.debug("converted event %s as %s", change, event)
+        for r in remove:
+            self._ids.pop(r, None)
 
-                yield event
+        if path:
+            self._ids[path] = oid
+
+        return event
+
+    def _filter_event(self, event):
+        # event filtering based on root path and event path
+
+        if not event:
+            return EventFilter.IGNORE
+
+        if not self._root_path:
+            return EventFilter.PROCESS
+
+        state_path = self.sync_state.get_path(event.oid)
+        prior_subpath = self.is_subpath_of_root(state_path)
+        if not event.exists:
+            # delete - ignore if not in state, or in state but is not subpath of root
+            return EventFilter.PROCESS if prior_subpath else EventFilter.IGNORE
+
+        if event.path:
+            curr_subpath = self.is_subpath_of_root(event.path)
+            if curr_subpath and not prior_subpath:
+                # Can't differentiate between creates and renames without watching the entire filesystem:
+                # Event has an oid and a current path, its a rename if the oid was seen before,
+                # but since events outside root are ignored we don't catch the case where an item is
+                # created outside root and then renamed into root.
+                # Hence the walk for directories -- a tradeoff for ignoring "outside root" events.
+                log.debug("created in or renamed into root: %s", event.path)
+                if event.otype == DIRECTORY:
+                    return EventFilter.WALK
```

Comment:
the main reason we needed this for OneDrive is that OneDrive generates a single event for folder renames (the event is for the folder being renamed, and no events come in for child files/folders)

does gdrive behave the same way?

Reply:
Yes, just double checked, gdrive behaves the sames way.